app/scripts/lib/WeakRefObjectMap.ts

```diff
@@ -0,0 +1,74 @@
+type WeakRefCompatibleObject<T> = {
+  [P in keyof T]: T[P] extends object ? WeakRef<T[P]> : never;
+};
+
+export class WeakRefObjectMap {
```

Comment:
Actually even that wouldn't be totally type compatible, because the `SelectedNetworkController` needs a `NetworkProxy` for the type. But this class doesn't need to know anything about that. So this would be a great use case for a generic type parameter.

e.g. something like:

```
class WeakRefObjectMap<WeakReffedObject extends object> implements Map<string, WeakReffedObject> {
```

Reply:
🤔 won't object suffice in that case? The object is not actually weakreffed its properties are?

Reply:
done here: https://github.com/MetaMask/metamask-extension/pull/23525/commits/94c4976528bb0f83c9815acd22be4a45fa4f5646

Reply:
It'll need to match the type here: https://github.com/MetaMask/core/blob/e86b84e6260beb689d95a4bed724d5a55237f35c/packages/selected-network-controller/src/SelectedNetworkController.ts#L101

That's why `object` wouldn't be specific enough; the controller expects it to be a map of `NetworkProxy` objects specifically

Reply:
ah I see ok!